storing-app-data/slides.md

```diff
@@ -0,0 +1,456 @@
+
+---
+
+# When working on an app "at scale", there are some extra issues you have to worry about.
+
+^ When you have tons of users, you'll see all of the edge cases you might normally brush off.
+
+---
+
+# [fit] Storage at Scale
+
+^ Let's talk about the concerns we have when storing data on disk, and then offer a solution that helps with these problems both at scale and in your hobby app.
+
+---
+
+# What is storage at scale?
+
+## Performing I/O operations on millions of devices, millions of times a day, globally.
+
+^ Being successful 99.9% of the time is no longer good enough. “Edge cases” become regular cases. Binaries live in the wild for months or years without getting updated, not all your customers are on flagship devices
+
+---
+
+---
+
+# [fit] Storage Reliability
+
+---
+
+# Why is storage reliability important?
+
+## The application should not crash due to a storage error
+
+---
+
+# Storage Reliability
+
+Application logic needs to be able to rely on callbacks being correct and timely.
+
+When a I/O failure happens, the application needs to ensure it performs the correct followup behavior.
+
+^ Storage reliability is about things actually being on disk when success is returned, and ensuring that if disk is latent or otherwise problematic, application logic can handle the issue without crashing or entering a bad state. Take for example storing if someone has accepted a specific version of your Terms of Service or a GDPR consent. If you use SharedPreferences#apply and the disk is full, you’ll never know that the write to disk failed. You’ll then lose the ability to run personalized advertising for that user, impacting revenue. With a reliable storage framework, you would get a callback that the write failed and why, giving you the option to prompt the user or block access from the app.
+
+---
+
+# Storage Reliability
+
+[.code-highlight: all]
+[.code-highlight: 4, 7]
+
+```java
+SharedPreferences sharedPref = context.getSharedPreferences("app", Context.MODE_PRIVATE);
+SharedPreferences.Editor editor = sharedPref.edit();
+editor.putString("auth_token", authToken);
+editor.apply();
+
+context.startActivity(MainScreenActivity.class);
+finish();
+```
+
+^ Here is example code for storing something via Shared Preferences. This is could that could very easily be part of your login process. But it has two bugs. If apply fails to write to disk, the user will be mysteriously logged out if you application is restarted. Additionally if the write is not immediate, the finish() will block the main thread on it completing.
+
+---
+
+## Even more important for legal and privacy status, such as a EULA or GDPR opt-out.
+
+^ Kicking your user out of the app due to an error is unpleasant, losing money is bad.
+
+---
+
+# [fit] Storage Performance
+
+---
+
+# Why is storage performance important?
+
+## The user experience should always be smooth.
+
+## Storage can cause frame drops.
+
+---
+
+# It is easy to drop frames
```

Comment:
Add some context on vsync and frame rendering, preferably with a timeline of 16ms blocks